_posts/2015-08-04-swift-ly-moving-forward.md

```diff
@@ -0,0 +1,225 @@
+to try if you don't know about [the XCPlayground framework](http://nshipster.com/xcplayground/) and the 
+`XCPSetExecutionShouldContinueIndefinitely` function. (Now you know.)
+
+A playground can show you the values of variables and more. Here you can see
+the previous summing example as it appears in the playground:
+
+![Tools](/img/swift-ly-moving-forward/swift-playground.png)
+
+## Functional Swift
+
+Here at Solita, functional programming is quite well established, with several
+Clojure projects in production. Therefore I've developed a special interest in 
+the functional aspects of Swift. It is by no means primarily a functional 
+programming language, but many of the key concepts, such as higher-order functions,
+programming with values, and closures, are well supported.
+
+One of the biggest differences with mainstream programming languages is the 
+pervasive use of immutable values in Swift. Indeed, you should take a page
+from the functional programming playbook and start using structs and `let` 
+as much as possible, instead
+of classes and `var`, since it helps to keep things straight in an increasingly
+multithreaded environment.
+
+The [Swift standard library](https://developer.apple.com/library/prerelease/ios/documentation/General/Reference/SwiftStandardLibraryReference/) 
+has a number of functions providing basic functional
+replacements for traditional loop-based iteration of array-like structures, 
+namely `map`, `filter` and `reduce`. By internalising these concepts you can 
+make your code more elegant and less cluttered.
+
+Here is an actual example of a mobile client application for rail traffic data, 
+filtering out trains by type from an array (slightly simplified from
+production code):
+
+    let filteredTrains = trains!.filter({ $0.trainCategory == "Commuter" || $0.trainCategory == "Long-distance" })
+
+In Objective-C, the standard way of filtering an array is traditionally
+quite verbose in comparison:
+
+    NSMutableArray *filteredTrains = [NSMutableArray array];
+    for (Train *train in trains) {
+        if (train.trainCategory == "Commuter" || train.trainCategory == "LongDistance") {
+            [filteredTrains addObject:train];
+        }
+    }
+
+For a functional way of summing the odd numbers like above, try this:
+
+    let series = [ 0, 1, 1, 2, 3, 5, 8, 13, 21,
+                   34, 55, 89, 144, 233, 377, 610 ]
+    var totalOdds = series
+        .filter { (term) in term % 2 == 0 }
+        .reduce(0) { (total, n) in total + n }
+    println(totalOdds)    
+    
+This looks quite incomprehensible at first, but if you break it down with the
+help of a playground, you can see that the array is filtered down to six elements,
+which are then summed together starting with zero. Of the three basic tools --
+`map`, `filter` and `reduce` -- the `reduce` function is most difficult to understand,
+but not much more than Quicksort, which also seems unintuitive at first.
+
+## Ready for prime time
```

Comment:
Siirtäisin tämän luvun ylöspäin Stuck with Objective-C -luvun edelle.  Vaikuttaa tärkeämmältä kuin historiikki tekstin idean kannalta, eikä jäsennys nähdäkseni menisi rikki.